vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,304 @@
+
+        Raises:
+            UnsupportedMetadataOperation if no implementation supports this operation.
+        """
+
+        if not has_copy_metadata():
+            MetadataHandler.raise_exception()
+
+        failed = []
+
+        for name, backend in _registry.items():
+            with contextlib.suppress(NotImplementedError):
+                if not backend(self._path).copy_metadata(dest, reset_orientation):
+                    failed.append(name)
+
+        if failed:
+            _logger.warning(
+                f"The following plugins failed to copy metadata: "
+                f"{', '.join(failed)}<br>\n"
+                f"Some metadata may be missing in the destination image {dest}."
+            )
+
+    def get_date_time(self) -> str:
+        """Get creation date and time as formatted string.
+
+        Raises:
+            UnsupportedMetadataOperation if no implementation supports this operation.
+        """
+
+        if not has_get_date_time():
+            MetadataHandler.raise_exception()
+
+        for _, backend in _registry.items():
+            with contextlib.suppress(NotImplementedError):
+                return backend(self._path).get_date_time()
+        return ""
+
+    def get_metadata(self, keys: Sequence[str]) -> MetadataDictT:
+        """Get value of all desired keys.
+
+        Use all registered implementations for `Operations.get_metadata` to query
+        the current image using the provided metadata keys. The result of all
+        implementations is combined.
+
+        Args:
+            keys: Keys of metadata to query the image for.
+
+        Returns:
+            Dictionary with retrieved metadata.
+
+        Raises:
+            UnsupportedMetadataOperation if no implementation supports this operation.
+        """
+
+        if not has_get_metadata():
+            MetadataHandler.raise_exception()
+
+        out: MetadataDictT = {}
+
+        for _, backend in _registry.items():
+            with contextlib.suppress(NotImplementedError):
+                # TODO: from 3.9 on use: c = a | b
+                out = {**backend(self._path).get_metadata(keys), **out}
+
+        return out
+
+    def get_keys(self) -> Iterable[str]:
+        """Retrieve the key of all metadata values available in the current image.
+
+        Raises:
+            UnsupportedMetadataOperation if no implementation supports this operation.
+        """
+
+        if not has_get_keys():
+            MetadataHandler.raise_exception()
+
+        out: Iterable[str] = iter([])
+
+        for _, backend in _registry.items():
+            with contextlib.suppress(NotImplementedError):
+                out = itertools.chain(backend(self._path).get_keys(), out)
+
+        return out
+
+    @staticmethod
+    def raise_exception() -> NoReturn:
+        """Raise an exception for a operations without implementation."""
+        msg = f"{inspect.stack()[1][3]} has no implementation"
```

Comment:
Random note flying over this, I would really avoid using the stack here :laughing: Maybe just call `raise_exception` with a string? Or, if you really want to avoid re-writing the name, decorate these methods with a decorator somewhere along:
```python
def metadata_function(check):
    def decorator(func):
        def inner(self, *args, **kwargs):
            if not check():
                self.raise_exception(func.__name__)
            return func(self, *args, **kwargs)
        return inner
    return decorator
```
but even this is probably over the top.

Reply:
fixed in aafe415fb7230ed7b63342eee41de7eeafdb190d